src/Mod/Assembly/App/AssemblyUtils.cpp

```diff
@@ -165,85 +468,376 @@ JointType getJointType(App::DocumentObject* joint)
+        if (std::next(it) == std::prev(names.end())) {
+            return obj;
+        }
+
+        if (obj->isDerivedFrom<App::Part>() || obj->isLinkGroup()) {
+            continue;
+        }
+        else if (obj->isDerivedFrom<PartDesign::Body>()) {
+            return handlePartDesignBody(obj, it);
+        }
+        else if (obj->isDerivedFrom<PartApp::Feature>()) {
+            // Primitive, fastener, gear, etc.
+            return obj;
+        }
+        else if (obj->isLink()) {
+            App::DocumentObject* linked_obj = obj->getLinkedObject();
+            if (linked_obj->isDerivedFrom<PartDesign::Body>()) {
+                auto* retObj = handlePartDesignBody(linked_obj, it);
+                return retObj == linked_obj ? obj : retObj;
+            }
+            else if (linked_obj->isDerivedFrom<PartApp::Feature>()) {
+                return obj;
+            }
+            else {
+                doc = linked_obj->getDocument();
+                continue;
+            }
+        }
+    }
+
+    return nullptr;
+}
+
+App::DocumentObject* getObjFromRef(App::PropertyXLinkSub* prop)
+{
+    if (!prop) {
         return nullptr;
     }
 
-    if (objName == containingPart->getNameInDocument()) {
-        return containingPart;
+    App::DocumentObject* obj = prop->getValue();
+    if (!obj) {
+        return nullptr;
     }
 
-    if (containingPart->getTypeId().isDerivedFrom(App::Link::getClassTypeId())) {
-        App::Link* link = dynamic_cast<App::Link*>(containingPart);
+    std::vector<std::string> subs = prop->getSubValues();
+    if (subs.empty()) {
+        return nullptr;
+    }
 
-        containingPart = link->getLinkedObject();
-        if (!containingPart) {
-            return nullptr;
+    return getObjFromRef(obj, subs[0]);
+}
+
+App::DocumentObject* getObjFromRef(App::DocumentObject* joint, const char* pName)
+{
+    if (!joint) {
+        return nullptr;
+    }
+
+    auto* prop = dynamic_cast<App::PropertyXLinkSub*>(joint->getPropertyByName(pName));
+
+    return getObjFromRef(prop);
+}
+
+App::DocumentObject* getMovingPartFromRef(App::DocumentObject* obj,
+                                          std::string& sub)
+{
+    if (!obj) {
+        return nullptr;
+    }
+
+    App::Document* doc = obj->getDocument();
+
+    std::vector<std::string> names = Base::Tools::splitSubName(sub);
+    names.insert(names.begin(), obj->getNameInDocument());
+
+    bool assemblyPassed = false;
+
+    for (const auto& objName : names) {
+        obj = doc->getObject(objName.c_str());
+        if (!obj) {
+            continue;
+        }
+
+        if (obj->isLink()) {  // update the document if necessary for next object
+            doc = obj->getLinkedObject()->getDocument();
+        }
+
+        if (obj->isDerivedFrom<AssemblyObject>()) {
+            // We make sure we pass the assembly for cases like part.assembly.part.body
+            assemblyPassed = true;
+            continue;
         }
```

Comment:
This if statement used to be `if (obj == this) {` like this:
```c++
        if (obj == this) {
            // We make sure we pass the assembly for cases like part.assembly.part.body
            assemblyPassed = true;
            continue;
        }
```
It needs to be verified, that this change doesn't break anything, before merging.